pennylane/circuit_graph.py

```diff
@@ -206,6 +211,124 @@ def hash(self):
         """
         return hash(self.serialize())
 
+    def to_openqasm(self, rotations=True):
+        """Serialize the circuit as an OpenQASM 2.0 program.
+
+        Only operations are serialized; all measurements
+        are assumed to take place in the computational basis.
+
+        .. note::
+
+            The serialized OpenQASM program assumes that gate definitions
+            in ``qelib1.inc`` are available.
+
+        Args:
+            rotations (bool): in addition to serializing user-specified
+                operations, also include the gates that diagonalize the
+                measured wires such that they are in the eigenbasis of the circuit observables.
+
+        Returns:
+            str: OpenQASM serialization of the circuit
+        """
+        # We import decompose_queue here to avoid a circular import
+        from pennylane.qnodes.base import decompose_queue  # pylint: disable=import-outside-toplevel
+
+        # add the QASM headers
+        qasm_str = "OPENQASM 2.0;\n"
+        qasm_str += 'include "qelib1.inc";\n'
+
+        if self.num_wires == 0:
+            # empty circuit
+            return qasm_str
+
+        # create the quantum and classical registers
+        qasm_str += "qreg q[{}];\n".format(self.num_wires)
+        qasm_str += "creg c[{}];\n".format(self.num_wires)
+
+        # Map PennyLane gate names to equivalent QASM gate names.
+        # Note that QASM has two native gates:
+        #
+        # - U (equivalent to qml.U3)
+        # - CX (equivalent to qml.CNOT)
+        #
+        # All other gates are defined in the file qelib1.inc:
+        # https://github.com/Qiskit/openqasm/blob/master/examples/generic/qelib1.inc
+        #
+        # Question: do other implementations of QASM exist that
+        # don't use the qelib1.inc gate definitions?
+        # Perhaps we allow developers to pass their own native_qasm_gates dictionary
+        # to this method, which is used instead of the default below
+        # if provided.
+        native_qasm_gates = {
```

Comment:
Nitpick: wondering if it would be worth moving this dictionary and the mock device class to the very beginning of the function? :thinking: (or even potentially outside of it) This would just create more focus on the logic inside

Reply:
That's a good idea!

Reply:
Done